src/target/riscv/riscv-013.c

```diff
@@ -3146,10 +3130,33 @@ static int modify_privilege(struct target *target, uint64_t *mstatus, uint64_t *
 	*mstatus = set_field(*mstatus, MSTATUS_MPRV, 1);
 
 	/* Write MSTATUS */
-	if (*mstatus == *mstatus_old)
+	if (*mstatus != *mstatus_old &&
+			register_write_direct(target, GDB_REGNO_MSTATUS, *mstatus) != ERROR_OK)
+		return ERROR_FAIL;
+
+	/* dcsr.mprven <- 1 */
+	return register_write_direct(target, GDB_REGNO_DCSR,
+			set_field(dcsr, CSR_DCSR_MPRVEN, CSR_DCSR_MPRVEN_ENABLED));
+}
+
+static int restore_privilege(struct target *target, uint64_t mstatus, uint64_t mstatus_old)
+{
+	if (!riscv_virt2phys_mode_is_hw(target))
 		return ERROR_OK;
 
-	return register_write_direct(target, GDB_REGNO_MSTATUS, *mstatus);
+	/* Restore MSTATUS */
+	if (mstatus != mstatus_old &&
+		    register_write_direct(target, GDB_REGNO_MSTATUS, mstatus_old) != ERROR_OK)
+		return ERROR_FAIL;
+
+	/* Read DCSR */
+	riscv_reg_t dcsr;
+	if (register_read_direct(target, &dcsr, GDB_REGNO_DCSR) != ERROR_OK)
+		return ERROR_FAIL;
+
+	/* dcsr.mprven <- 0 */
+	return register_write_direct(target, GDB_REGNO_DCSR,
```

Comment:
Please, use the operations that work with the cache (`riscv_reg_get/riscv_reg_set`).
There is no need to bypass the cache here.
Moreover, current implementation may introduce cache inconsistency. Consider the following:
```
> reg dcsr <some value with "mprven" set>
> reg dcsr # After this the "dcsr" value with "mprven" set is cached.
<some value with "mprven" set>
> mww <some addr> #After this "dcsr" on target has "mprven" cleared.
> reg dcsr
<some value with "mprven" set>
> reg dcsr -force
<some value with "mprven" cleared>
```
This is not a big deal since IMHO it's okay for the debugger to change "dcsr" however it likes but this is a peculiar behavior that may be easily avoided by using `riscv_reg_get/riscv_reg_set`.

As you have mentioned in #1188 `register_read/write_direct()` are used throughout `riscv-013.c`. However, most of the time this is not necessary and may introduce cache inconsistency. So I plan to phase out  `register_read/write_direct()`.

Reply:
address